simd/src/Kokkos_SIMD_Scalar.hpp

```diff
@@ -196,7 +196,7 @@ template <class T>
 template <class T>
 [[nodiscard]] KOKKOS_FORCEINLINE_FUNCTION simd<T, simd_abi::scalar> abs(
     simd<T, simd_abi::scalar> const& a) {
-  return simd<T, simd_abi::scalar>(std::abs(static_cast<T>(a)));
+  return simd<T, simd_abi::scalar>(Kokkos::abs<T>(a[0]));
```

Comment:
```suggestion
  return simd<T, simd_abi::scalar>(Kokkos::abs(a[0]));
```

Reply:
This gave
```
/__w/kokkos/kokkos/simd/src/Kokkos_SIMD_Scalar.hpp:199:47: error: call of overloaded 'abs(Kokkos::Experimental::simd<unsigned int, Kokkos::Experimental::simd_abi::scalar>::value_type)' is ambiguous
  199 |   return simd<T, simd_abi::scalar>(Kokkos::abs(a[0]));
      |                                    ~~~~~~~~~~~^~~~~~
In file included from /__w/kokkos/kokkos/core/src/Kokkos_Core.hpp:53,
                 from /__w/kokkos/kokkos/simd/src/Kokkos_SIMD_Common.hpp:23,
                 from /__w/kokkos/kokkos/simd/src/Kokkos_SIMD.hpp:20:
/__w/kokkos/kokkos/core/src/Kokkos_MathematicalFunctions.hpp:283:28: note: candidate: 'int Kokkos::abs(int)'
  283 | KOKKOS_INLINE_FUNCTION int abs(int n) {
      |                            ^~~
/__w/kokkos/kokkos/core/src/Kokkos_MathematicalFunctions.hpp:287:29: note: candidate: 'long int Kokkos::abs(long int)'
  287 | KOKKOS_INLINE_FUNCTION long abs(long n) {
      |                             ^~~
/__w/kokkos/kokkos/core/src/Kokkos_MathematicalFunctions.hpp:296:34: note: candidate: 'long long int Kokkos::abs(long long int)'
  296 | KOKKOS_INLINE_FUNCTION long long abs(long long n) {
      |                                  ^~~
/__w/kokkos/kokkos/core/src/Kokkos_MathematicalFunctions.hpp:305:30: note: candidate: 'float Kokkos::abs(float)'
  305 | KOKKOS_INLINE_FUNCTION float abs(float x) {
      |                              ^~~
/__w/kokkos/kokkos/core/src/Kokkos_MathematicalFunctions.hpp:309:31: note: candidate: 'double Kokkos::abs(double)'
  309 | KOKKOS_INLINE_FUNCTION double abs(double x) {
      |                               ^~~
/__w/kokkos/kokkos/core/src/Kokkos_MathematicalFunctions.hpp:313:20: note: candidate: 'long double Kokkos::abs(long double)'
  313 | inline long double abs(long double x) {
      |                    ^~~

Reply:
But I don't even understand how `Kokkos::abs<T>` worked since it is not a function template.

Reply:
Holly s, was it going through the complex overload?!!

Reply:
Yup
```diff
diff --git a/core/src/Kokkos_Complex.hpp b/core/src/Kokkos_Complex.hpp
index 4cecbe4cf..40ac102a4 100644
--- a/core/src/Kokkos_Complex.hpp
+++ b/core/src/Kokkos_Complex.hpp
@@ -44,6 +44,8 @@ class
     alignas(2 * sizeof(RealType))
 #endif
         complex {
+  static_assert(std::is_floating_point_v<RealType>);
+
  private:
   RealType re_{};
   RealType im_{};
```
yields
```
In file included from /Users/qdi/CLionProjects/kokkos/simd/unit_tests/TestSIMD.cpp:19:
In file included from /Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD.hpp:20:
In file included from /Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_Common.hpp:23:
In file included from /Users/qdi/CLionProjects/kokkos/core/src/Kokkos_Core.hpp:54:
In file included from /Users/qdi/CLionProjects/kokkos/core/src/Kokkos_MathematicalSpecialFunctions.hpp:31:
/Users/qdi/CLionProjects/kokkos/core/src/Kokkos_Complex.hpp:47:3: error: static_assert failed due to requirement 'std::is_floating_point_v<unsigned long long>'
  static_assert(std::is_floating_point_v<RealType>);
  ^             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_Scalar.hpp:199:51: note: in instantiation of template class 'Kokkos::complex<unsigned long long>' requested here
  return simd<T, simd_abi::scalar>(Kokkos::abs<T>(a[0]));
                                                  ^
/Users/qdi/CLionProjects/kokkos/simd/src/Kokkos_SIMD_Scalar.hpp:199:36: note: while substituting deduced template arguments into function template 'abs' [with RealType = unsigned long long]
  return simd<T, simd_abi::scalar>(Kokkos::abs<T>(a[0]));
                                   ^

Reply:
Try
```C++
if constexpr (std::is_signed_v<T>) {
  return (a < 0 ? -a : a);
} else {
  return a;
}

Reply:
Fixed. Removed `else` statement to avoid that `missing return statement` warnings on cuda build.